cirq-core/cirq/work/zeros_sampler.py

```diff
@@ -53,17 +54,31 @@ def run_sweep(
             resolver.
 
         Raises:
-            ValueError if this sampler has a device and the circuit is not
-            valid for the device.
+            ValueError: circuit is not valid for the sampler, due to invalid
+            repeated keys or incompatibility with the sampler's device.
         """
         if self.device:
             self.device.validate_circuit(program)
-        measurements: Dict[str, np.ndarray] = {}
+        num_qubits: Dict[str, int] = {}
+        num_instances: Dict[str, int] = collections.Counter()
         for op in program.all_operations():
             key = protocols.measurement_key_name(op, default=None)
             if key is not None:
-                measurements[key] = np.zeros((repetitions, len(op.qubits)), dtype=int)
+                n = len(op.qubits)
+                prev_n = num_qubits.setdefault(key, n)
+                if n != prev_n:
+                    raise ValueError(
+                        "Different num qubits for repeated measurement: "
```

Comment:
Two things:
- This also needs to check that the dimensions for each qudit are the same - e.g. (3, 2) != (2, 3).
- This feels like the wrong place for this check to live, since presumably every sampler/simulator is going to require it. Could we move it up to the Sampler superclass?

Reply:
Moved this into a helper function `Sampler._get_measurement_shapes` in the base class. Note that this is basically a preliminary version of the "measurement_key_shape" protocol proposed in https://github.com/quantumlib/Cirq/issues/4888.